In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
path = r'C:\Users\stefa\Documents\07-Instacart Basket Analysis'

In [3]:
path

'C:\\Users\\stefa\\Documents\\07-Instacart Basket Analysis'

In [4]:
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'),  index_col = False)

In [5]:
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'),  index_col = False)

In [6]:
df_ords = df_ords.drop(columns = ['Unnamed: 0'])
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       orders_time_of_day (24 hours)  days_since_prior_order  
count                   3.421083e+06            3.214874e+06  
mean                    1.345202e+01            1.111484e+01  
std                     4.226088e+00            9.206737e+00  
min                     0.000000e+00            0.000000e+00  
25%                     1.000000e+01            4.000000e+00  
50%                     1.300000e+01            7.000000e+00  
75%                     1.600000e+01            1.500000e+01  
max                     2.300000e+01            3.000000e+01

In [7]:
# Create a dataframe
df_test = pd.DataFrame()

Finding mixed data types

In [8]:
# Create a mixed test column
df_test['mix'] = ['a', 'b', 1, True]

In [9]:
# Display head of df_test dataframe
df_test.head()

mix
0     a
1     b
2     1
3  True

In [10]:
# Check for mixed types

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


C:\Users\stefa\AppData\Local\Temp\ipykernel_21568\1522843687.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [11]:
# Change variable type from int64 to string
df_test['mix'] = df_test['mix'].astype('str')

Finding missing values

In [12]:
# Finding missing values in df_prods dataframe
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [13]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [14]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [15]:
# Finding the shape of df_prods dataframe
df_prods.shape

(49693, 5)

In [16]:
# Creating new dataframe df_prods_clean
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [17]:
# Finding the shape of df_prods_clean dataframe
df_prods_clean.shape

(49677, 5)

Finding duclicate data

In [18]:
# Looking for duplicates in df_prods_clean dataframe
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [19]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [20]:
## Creating new dataframe df_prods_clean_no_dups
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [21]:
df_prods_clean_no_dups.shape

(49672, 5)

In [22]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

All previous code was for practicing applying data consistency checks using the df_prods data set. The code below is for completing task 4.5

In [23]:
# Step 2
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       orders_time_of_day (24 hours)  days_since_prior_order  
count                   3.421083e+06            3.214874e+06  
mean                    1.345202e+01            1.111484e+01  
std                     4.226088e+00            9.206737e+00  
min                     0.000000e+00            0.000000e+00  
25%                     1.000000e+01            4.000000e+00  
50%                     1.300000e+01            7.000000e+00  
75%                     1.600000e+01            1.500000e+01  
max                     2.300000e+01            3.000000e+01

I can not find any anomalies with the df_ords dataframe.

In [24]:
# Step 3
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)      

In [25]:
# Step 4

There are no mixed data types in the df_ords dataset.

In [26]:
# Step 5
df_ords.isnull().sum()

order_id                              0
user_id                               0
order_number                          0
orders_day_of_week                    0
orders_time_of_day (24 hours)         0
days_since_prior_order           206209
dtype: int64

In [27]:
df_ords.shape

(3421083, 6)

There are 206209 null entries in the df_orbs dataframe. These null entries are all in the days_since_prior_order column. The reason for these missing entries I believe is that those enteries were input incorrectly, missing the information about how long it had been since a customer ordered previously. Therefore, we are seeing gaps in our dataframe.

In [28]:
# Step 6
df_mis = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [29]:
df_mis

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
11        2168274        2             1                   2   
26        1374495        3             1                   1   
39        3343014        4             1                   6   
45        2717275        5             1                   3   
...           ...      ...           ...                 ...   
3420930    969311   206205             1                   4   
3420934   3189322   206206             1                   3   
3421002   2166133   206207             1                   6   
3421019   2227043   206208             1                   1   
3421069   3154581   206209             1                   3   

         orders_time_of_day (24 hours)  days_since_prior_order  
0                                    8                     NaN  
11                                  11                     NaN  
26                                  14                     NaN  
39                                  11                     NaN  
45                                  12                     NaN  
...                                ...                     ...  
3420930                             12                     NaN  
3420934                             18                     NaN  
3421002                             19                     NaN  
3421019                             15                     NaN  
3421069                             11                     NaN  

[206209 rows x 6 columns]

In [34]:
df_mis.head(30)

order_id  user_id  order_number  orders_day_of_week  \
0     2539329        1             1                   2   
11    2168274        2             1                   2   
26    1374495        3             1                   1   
39    3343014        4             1                   6   
45    2717275        5             1                   3   
50    2086598        6             1                   5   
54    2565571        7             1                   3   
75     600894        8             1                   6   
79     280530        9             1                   1   
83    1224907       10             1                   2   
89    2411567       11             1                   0   
97    1483446       12             1                   1   
103   2618231       13             1                   6   
116   2906263       14             1                   2   
130   2514324       15             1                   5   
153   1273676       16             1                   2   
160   1737705       17             1                   2   
201   2780889       18             1                   6   
208   1321974       19             1                   6   
218    947722       20             1                   1   
223   2625565       21             1                   1   
257   2300537       22             1                   3   
273   1792521       23             1                   5   
278    611214       24             1                   3   
297   2610576       25             1                   1   
301   3229714       26             1                   2   
314    361493       27             1                   3   
396   1138968       28             1                   0   
421   2064090       29             1                   5   
440    375303       30             1                   1   

     orders_time_of_day (24 hours)  days_since_prior_order  
0                                8                     NaN  
11                              11                     NaN  
26                              14                     NaN  
39                              11                     NaN  
45                              12                     NaN  
50                              18                     NaN  
54                               9                     NaN  
75                               0                     NaN  
79                              17                     NaN  
83                              14                     NaN  
89                              11                     NaN  
97                              12                     NaN  
103                             12                     NaN  
116                             18                     NaN  
130                             13                     NaN  
153                              8                     NaN  
160                             13                     NaN  
201                             18                     NaN  
208                             13                     NaN  
218                             10                     NaN  
223                             10                     NaN  
257                              7                     NaN  
273                             14                     NaN  
278                             20                     NaN  
297                             14                     NaN  
301                             15                     NaN  
314                              9                     NaN  
396                             14                     NaN  
421                             11                     NaN  
440                             13                     NaN

In [35]:
df_mis.tail(30)

order_id  user_id  order_number  orders_day_of_week  \
3420548    566068   206180             1                   0   
3420567    351920   206181             1                   0   
3420582    485448   206182             1                   0   
3420593     31169   206183             1                   0   
3420608    180109   206184             1                   6   
3420613   2168901   206185             1                   5   
3420624   1831589   206186             1                   6   
3420628    474057   206187             1                   2   
3420663   1944921   206188             1                   4   
3420671   2966736   206189             1                   4   
3420678   1955107   206190             1                   6   
3420687    936117   206191             1                   6   
3420693   3242108   206192             1                   6   
3420708    313609   206193             1                   5   
3420750   3395125   206194             1                   1   
3420762    926652   206195             1                   3   
3420782    825502   206196             1                   6   
3420787   3325493   206197             1                   2   
3420811   2154947   206198             1                   0   
3420819   3302466   206199             1                   2   
3420839   2508134   206200             1                   5   
3420863     28210   206201             1                   6   
3420896    799173   206202             1                   2   
3420919   1991850   206203             1                   1   
3420925   1438269   206204             1                   1   
3420930    969311   206205             1                   4   
3420934   3189322   206206             1                   3   
3421002   2166133   206207             1                   6   
3421019   2227043   206208             1                   1   
3421069   3154581   206209             1                   3   

         orders_time_of_day (24 hours)  days_since_prior_order  
3420548                             11                     NaN  
3420567                             15                     NaN  
3420582                              8                     NaN  
3420593                             13                     NaN  
3420608                             12                     NaN  
3420613                             19                     NaN  
3420624                             14                     NaN  
3420628                             23                     NaN  
3420663                             10                     NaN  
3420671                             15                     NaN  
3420678                              9                     NaN  
3420687                             15                     NaN  
3420693                              6                     NaN  
3420708                             15                     NaN  
3420750                              9                     NaN  
3420762                             16                     NaN  
3420782                             14                     NaN  
3420787                             19                     NaN  
3420811                             16                     NaN  
3420819                             11                     NaN  
3420839                             19                     NaN  
3420863                             14                     NaN  
3420896                             12                     NaN  
3420919                             13                     NaN  
3420925                             11                     NaN  
3420930                             12                     NaN  
3420934                             18                     NaN  
3421002                             19                     NaN  
3421019                             15                     NaN  
3421069                             11                     NaN

There is no need to replace the NaN values in the days_since_prior_order column of the df_ords dataframe. I was incorrect in my reasoning to why the days_since_prior_order column has NaN values. The data shows the rows that have NaN data in their days_since_prior_order column are first time orders, meaning there is no prior date to pull from.

In [37]:
# Step 7
df_dups_ords = df_ords[df_ords.duplicated()]

In [38]:
df_dups_ords

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, orders_time_of_day (24 hours), days_since_prior_order]
Index: []

In [39]:
# Step 8

There seem to be no duplicates in the df_ords data frame.

In [42]:
# Step 9
df_ords

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
1         2398795        1             2                   3   
2          473747        1             3                   3   
3         2254736        1             4                   4   
4          431534        1             5                   4   
...           ...      ...           ...                 ...   
3421078   2266710   206209            10                   5   
3421079   1854736   206209            11                   4   
3421080    626363   206209            12                   1   
3421081   2977660   206209            13                   1   
3421082    272231   206209            14                   6   

         orders_time_of_day (24 hours)  days_since_prior_order  
0                                    8                     NaN  
1                                    7                    15.0  
2                                   12                    21.0  
3                                    7                    29.0  
4                                   15                    28.0  
...                                ...                     ...  
3421078                             18                    29.0  
3421079                             10                    30.0  
3421080                             12                    18.0  
3421081                             12                     7.0  
3421082                             14                    30.0  

[3421083 rows x 6 columns]

In [43]:
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))

In [44]:
df_prods.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))